src/Elastic.Apm/Logging/TraceLogger.cs

```diff
@@ -0,0 +1,88 @@
+// Licensed to Elasticsearch B.V under
+// one or more agreements.
+// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information
+
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using Elastic.Apm.Helpers;
+
+namespace Elastic.Apm.Logging
+{
+	/// <summary>
+	/// A logging implementation that logs to a <see cref="TraceSource"/> with the source name <c>Elastic.Apm</c>
+	/// </summary>
+	internal class TraceLogger : IApmLogger, ILogLevelSwitchable
+	{
+		private const string SourceName = "Elastic.Apm";
+
+		private static readonly TraceSource TraceSource;
+
+		static TraceLogger() => TraceSource = new TraceSource(SourceName);
+
+		public TraceLogger(LogLevel level) => LogLevelSwitch = new LogLevelSwitch(level);
+
+		public LogLevelSwitch LogLevelSwitch { get; }
+
+		private LogLevel Level => LogLevelSwitch.Level;
+
+		public bool IsEnabled(LogLevel level) => Level <= level;
+
+		public void Log<TState>(LogLevel level, TState state, Exception e, Func<TState, Exception, string> formatter)
+		{
+			if (!IsEnabled(level)) return;
+
+			var message = formatter(state, e);
+
+			// default message size is 51 + length of loglevel (max 8), message and exception.

```

Comment:
I don't really get this comment. What do you mean by "default message size is 51"?

Looking at the code I'd think that the strategy here is to start with a fairly small number for the capacity, but then ultimately the capacity of the `StringBuilder` will grow based on the length of the messages we print. So in this case we start with `80` (line below) and then if the message is longer then `StringBuilder` will increase the capacity and once we reuse it for a later massage the capacity will be bigger. So it'll adapt to the longest message. Do I understand this correctly? 

What I wanna say here is that I don't really get the comment here :)

Reply:
It means that the complete log message size _at a minimum_ will be 51 characters, plus the variable length of log level (max 8 characters), message and exception. So any starting capacity should _at a minimum_ accommodate this, which the starting capacity of 80 characters will - assuming no exception and max log level length, leaves 21 characters for the message. This may be overly conservative, in which case a better estimate may be to use 59 + the calculated mean length of a log message.

`StringBuilder` will grow (and allocate) to accommodate larger log messages. The intention is to avoid allocating whilst serving the common usage.

Reply:
Got it - the term `default message size` was strange to me, but I got it.